src/main/scala/org/constellation/domain/redownload/RedownloadService.scala

```diff
@@ -1,38 +1,159 @@
-      (updated, ())
-    }
+  def persistLocalSnapshot(recentSnapshot: RecentSnapshot) = updateProposedSnapshots(Seq((cluster.id, recentSnapshot)))
 
-  def getOwnSnapshots(): F[Map[Long, String]] = ownSnapshots.get
+  def getLocalSnapshots() = proposedSnapshots.get.map { snapshots =>
+    snapshots.values.flatMap { proposalsForHeight =>
+      proposalsForHeight.get(cluster.id)
+    }.toSeq
+  }
 
-  def getOwnSnapshot(height: Long): F[Option[String]] =
-    ownSnapshots.get.map(_.get(height))
+  def getLocalSnapshotAtHeight(height: Long): F[Option[RecentSnapshot]] = proposedSnapshots.get.map { proposedSnaps =>
+    val snapshotsAtHeight = proposedSnaps.getOrElse(height, Map())
+    snapshotsAtHeight.get(cluster.id)
+  }
 
-  // TODO: Use cluster peers to fetch "/snapshot/own" and merge with peersProposals Map
-  def fetchPeersProposals(): F[Unit] = ???
+  private[redownload] def fetchPeerProposals(): F[List[(Id, RecentSnapshot)]] =
+    for {
+      peers <- cluster.readyPeers
+      chunkedPeerProposals <- peers.toList.traverse {
+        case (id, peerData) =>
+          val resp = peerData.client.getNonBlockingFLogged[F, Array[Array[Byte]]](
+            "snapshot/own",
+            timeout = 3.seconds,
+            tag = "snapshot/own"
+          )(C)
+          resp.map((id, _)).handleErrorWith { ex =>
+            Sync[F].delay(logger.error(s"fetchPeerProposals error - ${ex}")) >> Sync[F].delay((id, Array()))
+          }
+      }
+      deSerializedPeerProposals = chunkedPeerProposals.map(RedownloadService.deserializeProposals).flatMap {
+        case (id, recentSnaps) => recentSnaps.map(snap => (id, snap))
+      }
+    } yield deSerializedPeerProposals
+
+  private[redownload] def updateProposedSnapshots(fetchedProposals: Seq[(Id, RecentSnapshot)]) =
+    proposedSnapshots.modify { m =>
+      val updatedProposals = fetchedProposals.foldLeft(m) {
+        case (prevProposals, (id, recentSnap)) =>
+          val proposalsAtHeight = prevProposals.getOrElse(recentSnap.height, Map())
+          val hasMadeProposal = proposalsAtHeight.get(id).isDefined //todo Observation of new proposal
+          val test = prevProposals.updated(recentSnap.height, proposalsAtHeight + (id -> recentSnap))
+          if (!hasMadeProposal) test
+          else prevProposals
+      }
+      (updatedProposals, ())
+    }
 
-  // TODO: Check for the majority snapshot and store under a variable
-  def recalculateMajoritySnapshot(): F[Unit] = ???
+  def fetchAndSetPeerProposals() = fetchPeerProposals().flatMap(updateProposedSnapshots)
 
-  // TODO: Check for the alignment ownSnapshots <-> majoritySnapshot and call redownload if needed
-  def checkForAlignmentWithMajoritySnapshot(): F[Unit] = ???
+  def recalculateMajoritySnapshot(): F[(Seq[RecentSnapshot], Set[Id])] =
+    for {
+      allProposals <- proposedSnapshots.get
+      allProposalsNormalized = allProposals.toSeq.flatMap {
+        case (height, proposals: Map[Id, RecentSnapshot]) =>
```

Comment:
why you need to specify a type during the pattern matching?
I think there are no other proposals of different type